fhir-cql-server/src/main/java/com/ibm/fhir/cql/engine/server/retrieve/ServerFhirRetrieveProvider.java

```diff
@@ -0,0 +1,148 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+
+import com.ibm.fhir.cql.engine.retrieve.SearchParameterFhirRetrieveProvider;
+import com.ibm.fhir.cql.engine.retrieve.SearchParameterMap;
+import com.ibm.fhir.cql.engine.searchparam.IQueryParameter;
+import com.ibm.fhir.cql.engine.searchparam.ReferenceParameter;
+import com.ibm.fhir.cql.engine.searchparam.SearchParameterResolver;
+import com.ibm.fhir.cql.helpers.FhirBundleCursor;
+import com.ibm.fhir.model.resource.Bundle;
+import com.ibm.fhir.model.resource.Resource;
+import com.ibm.fhir.persistence.SingleResourceResult;
+import com.ibm.fhir.search.SearchConstants;
+import com.ibm.fhir.server.operation.spi.FHIRResourceHelpers;
+
+public class ServerFhirRetrieveProvider extends SearchParameterFhirRetrieveProvider {
+
+    private static final String DUMMY_REQUEST_URI = "https://localhost/fhir-server/api/v4";
+
+    private static Logger logger = Logger.getLogger(ServerFhirRetrieveProvider.class.getName());
+    
+    private FHIRResourceHelpers resourceHelpers;
+
+    public ServerFhirRetrieveProvider(FHIRResourceHelpers resourceHelpers, SearchParameterResolver searchParameterResolver) {
+        super(searchParameterResolver);
+        this.resourceHelpers = resourceHelpers;
+    }
+
+    @Override
+    protected Iterable<Object> executeQueries(String dataType, List<SearchParameterMap> queries) throws Exception {
+        List<Object> results = new ArrayList<>();
+
+        for (SearchParameterMap map : queries) {
+            if( logger.isLoggable(Level.FINE) ) {
+                logger.fine(String.format("Executing query %s", map.toString()));
+            }
+            
+            MultivaluedMap<String, String> queryParameters = getQueryParameters(map);
+
+            // _total=none instructs the server to skip the count(*) query which improves performance
+            queryParameters.putSingle(SearchConstants.TOTAL, "none");
+
+            Resource resource = executeQuery(dataType, queryParameters);
+            if (resource instanceof Bundle) {
+                AtomicInteger pageNumber = new AtomicInteger(1);
+                FhirBundleCursor cursor = new FhirBundleCursor(url -> {
+                    try {
+                        int nextPage = pageNumber.incrementAndGet();
+                        if( logger.isLoggable(Level.FINE) ) {
+                            logger.fine(String.format("Retrieving page %d / %s", nextPage, url));
+                        }
+                        queryParameters.putSingle(SearchConstants.PAGE, String.valueOf(nextPage));
+                        return resourceHelpers.doSearch(dataType, /*compartment=*/null, /*compartmentId=*/null, queryParameters, DUMMY_REQUEST_URI, /*contextResource=*/null);
```

Comment:
```suggestion
                        return resourceHelpers.doSearch(dataType, /*compartment=*/null, /*compartmentId=*/null, queryParameters, DUMMY_REQUEST_URI, /*contextResource=*/null);
```
@lmsurpre thoughts on using a DUMMY_REQUEST_URI here? any problems with it?